test-case/check-volume-levels.sh

```diff
@@ -0,0 +1,394 @@
+    wait
+
+    cset_unmute
+    wait
+
+    cset_mutevol
+    wait
+
+    cset_min
+    wait
+
+    cset_volume_diff2 "$CAP_CHANNELS"
+
+    wait_for_pid $PID
+    dlogi Ready.
+}
+
+# 0..1s channels 1-4 previous gains
+# 1..2s channels 1-4 different mute switches
+# 2..3s channels 1-4 all muted
+run_test_2 () {
+    arecord_in_background "$ARECORD_WAV2" "$CAP_CHANNELS" 4
+    PID=$?
+    wait
+
+    cset_nom
+    cset_mute_diff1 "$CAP_CHANNELS"
+    wait
+
+    cset_mute
+    wait
+
+    wait_for_pid $PID
+    dlogi Ready.
+}
+
+# 0..1s channels 1-4 muted
+# 1..2s channels 1-4  nominal gain
+# 2..3s different mute switches
+# 3..4s channels 1-4 nominal gain
+run_test_3 () {
+    arecord_in_background "$ARECORD_WAV3" "$CAP_CHANNELS" 4
+    PID=$?
+    wait
+
+    cset_nom
+    wait
+
+    cset_mute_diff2 "$CAP_CHANNELS"
+    wait
+
+    cset_unmute
+
+    wait_for_pid "$PID"
+    dlogi Ready.
+}
+
+#
+# Helper functions
+#
+
+cset_nom () {
+    dlogi Set nominal volume
+    amixer cset name="$CAP_VOLUME" $VOL_NOM
+}
+
+cset_max () {
+    dlogi Set maximum volume
+    amixer cset name="$CAP_VOLUME" $VOL_MAX
+}
+
+cset_min () {
+    dlogi Set minimum volume
+    amixer cset name="$CAP_VOLUME" $VOL_MIN
+}
+
+cset_mutevol () {
+    dlogi Set mute via volume
+    amixer cset name="$CAP_VOLUME" $VOL_MUTE
+}
+cset_mute () {
+    dlogi Set mute via switch
+    amixer cset name="$CAP_SWITCH" off
+}
+
+cset_unmute () {
+    dlogi Set unmute
+    amixer cset name="$CAP_SWITCH" on
+}
+
+wait () {
+    dlogi Waiting
+    sleep 1
+}
+
+cset_mute_diff1 () {
+    dlogi Set switch pattern 1
+    case "$1" in
+	4)
+	    amixer cset name="$CAP_SWITCH" off,on,on,off
+	    ;;
+	2)
+	    amixer cset name="$CAP_SWITCH" off,on
+	    ;;
+	1)
+	    amixer cset name="$CAP_SWITCH" off
+	    ;;
+    esac
+}
+
+cset_mute_diff2 () {
+    dlogi Set switch pattern 2
+    case "$1" in
+	4)
+	    amixer cset name="$CAP_SWITCH" on,off,off,on
+	    ;;
+	2)
+	    amixer cset name="$CAP_SWITCH" on,off
+	    ;;
+	1)
+	    amixer cset name="$CAP_SWITCH" on
+	    ;;
+    esac
+}
+
+cset_volume_diff1 () {
+    dlogi Set volume pattern 1
+    case "$1" in
+	4)
+	    amixer cset name="$CAP_VOLUME" $VOL_P10DB,$VOL_0DB,$VOL_M10DB,$VOL_M30DB
+	    ;;
+	2)
+	    amixer cset name="$CAP_VOLUME" $VOL_P10DB,$VOL_0DB
+	    ;;
+	1)
+	    amixer cset name="$CAP_VOLUME" $VOL_P10DB
+	    ;;
+    esac
+}
+
+cset_volume_diff2 () {
+    dlogi Set volume pattern 2
+    case "$1" in
+	4)
+	    amixer cset name="$CAP_VOLUME" $VOL_M10DB,$VOL_P10DB,$VOL_0DB,$VOL_M20DB
+	    ;;
+	2)
+	    amixer cset name="$CAP_VOLUME" $VOL_M10DB,$VOL_P10DB
+	    ;;
+	1)
+	    amixer cset name="$CAP_VOLUME" $VOL_M10DB
+	    ;;
+    esac
+}
+
+wait_for_pid () {
+    dlogi Waiting for process "$1" to complete
+    tail --pid="$pid" -f /dev/null
+}
+
+arecord_in_background () {
+    dlogi Capturing file "$1"
+    arecord -D "$CAP_HW" -f $CAP_FORMAT -r $CAP_RATE -c "$2" -d "$3" "$1"  2> /dev/null &
+    pid=$!
+    return $pid
```

Comment:
Hardcoding the background decision in the function and returning the PID is unusual. I think it's much clearer to have the `&` and the `wait` at the same place:

```shell
my_arecord_function &  arecPID=$?
...
...
wait $arecPID

Reply:
I think that I could make such work by exporting aplayPID and arecPID.

Reply:
I didn't manage to collect the PID correctly with set -e. There's now `wait %2` in use. It works if there's the aplay in background as %1. I haven't tested killing it during the test but I'm pretty confident the levels checker will return fail if there's no signal for volume. But I'll keep this open until confirmed the script doesn't give false pass.